<a href="https://colab.research.google.com/github/buke2016/buke2016/blob/buke2016-patch-3/hw_1_cmsi_533.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
#1 import numpy
import numpy as np

In [ ]:
#2 link to the alt_fuel_stations_Jan 2024.csv
df = pd.read_csv('C:\\Users\\bukep\\Spring 2024\\533_Data science\\Homework1\\alt_fuel_stations_Jan_14_2024.csv')

C:\Users\bukep\AppData\Local\Temp\ipykernel_31468\2954538210.py:2: DtypeWarning: Columns (6,16,20,31,33,39,40,41,43,46,52,55,57,58,60,62,65,67,69,71) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('C:\\Users\\bukep\\Spring 2024\\533_Data science\\Homework1\\alt_fuel_stations_Jan_14_2024.csv')


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94369 entries, 0 to 94368
Data columns (total 74 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   Fuel Type Code                           94369 non-null  object 
 1   Station Name                             94368 non-null  object 
 2   Street Address                           94335 non-null  object 
 3   Intersection Directions                  5089 non-null   object 
 4   City                                     94365 non-null  object 
 5   State                                    94354 non-null  object 
 6   ZIP                                      94369 non-null  object 
 7   Plus4                                    0 non-null      float64
 8   Station Phone                            87172 non-null  object 
 9   Status Code                              94369 non-null  object 
 10  Expected Date                            6521 

In [ ]:
#3 df.columns
# df_miss.head(3) for


In [ ]:
df.columns

Index(['Fuel Type Code', 'Station Name', 'Street Address',
       'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser N

In [ ]:
df.describe()

Plus4  EV Level1 EVSE Num  EV Level2 EVSE Num  EV DC Fast Count  \
count    0.0          755.000000        71367.000000      12062.000000   
mean     NaN            4.200000            2.410736          3.789753   
std      NaN            8.169848            3.162759          4.697930   
min      NaN            1.000000            1.000000          1.000000   
25%      NaN            1.000000            2.000000          1.000000   
50%      NaN            2.000000            2.000000          2.000000   
75%      NaN            3.500000            2.000000          4.000000   
max      NaN           90.000000          338.000000         84.000000   

           Latitude     Longitude             ID  Federal Agency ID  \
count  94369.000000  94369.000000   94369.000000        1322.000000   
mean      39.165938    -95.183994  187443.964406          13.563540   
std        5.711686     19.411451   78795.768102           6.158342   
min      -41.354100   -164.848855      17.000000           2.000000   
25%       34.392261   -117.574477  144203.000000           8.000000   
50%       39.399690    -89.316809  187321.000000          14.000000   
75%       42.975710    -78.593763  238224.000000          16.000000   
max       64.852466    122.373276  325442.000000          33.000000   

       CNG Dispenser Num  CNG Total Compression Capacity  \
count        1029.000000                      622.000000   
mean            3.029155                      915.500000   
std             6.505245                     1075.692844   
min             0.000000                        2.000000   
25%             1.000000                      250.000000   
50%             2.000000                      700.000000   
75%             2.000000                     1200.000000   
max            80.000000                    10620.000000   

       CNG Storage Capacity  RD Blends (French)  RD Maximum Biodiesel Level  
count            319.000000                 0.0                  537.000000  
mean           46889.423197                 NaN                    5.370577  
std            56246.507440                 NaN                    4.391574  
min                0.000000                 NaN                    5.000000  
25%            29203.500000                 NaN                    5.000000  
50%            36000.000000                 NaN                    5.000000  
75%            58000.000000                 NaN                    5.000000  
max           593136.000000                 NaN                   99.000000

In [ ]:
df['Fuel Type Code'].value_counts().sort_values(ascending=False)

Fuel Type Code
ELEC    82214
E85      4527
LPG      3528
BD       1702
CNG      1530
RD        610
HY        133
LNG       125
Name: count, dtype: int64

In [ ]:
df['Fuel Type Code'].describe()

count     94369
unique        8
top        ELEC
freq      82214
Name: Fuel Type Code, dtype: object

In [ ]:
fuel_type_codes = df['Fuel Type Code'].value_counts(normalize=True)

In [ ]:
num_unique_values = len(fuel_type_codes)

In [ ]:
percentage_values = (fuel_type_codes * 100).round(2)

In [ ]:
print("Number of unique values in 'Fuel Type Code':", num_unique_values)
print("\nPercentage of each value in 'Fuel Type Code':")
for code, percentage in zip(fuel_type_codes.index, percentage_values):
    print(f"{code}: {percentage}%")

Number of unique values in 'Fuel Type Code': 8

Percentage of each value in 'Fuel Type Code':
ELEC: 87.12%
E85: 4.8%
LPG: 3.74%
BD: 1.8%
CNG: 1.62%
RD: 0.65%
HY: 0.14%
LNG: 0.13%


In [ ]:
df.head(3)

Fuel Type Code                                  Station Name  \
0            CNG          Spire - Montgomery Operations Center   
1            CNG  Metropolitan Atlanta Rapid Transit Authority   
2            CNG                         United Parcel Service   

         Street Address Intersection Directions        City State    ZIP  \
0      2951 Chestnut St                     NaN  Montgomery    AL  36107   
1   2424 Piedmont Rd NE                     NaN     Atlanta    GA  30324   
2  270 Marvin Miller Dr                     NaN     Atlanta    GA  30336   

   Plus4 Station Phone Status Code  ... Restricted Access RD Blends  \
0    NaN           NaN           E  ...               NaN       NaN   
1    NaN           NaN           E  ...               NaN       NaN   
2    NaN           NaN           E  ...               NaN       NaN   

  RD Blends (French) RD Blended with Biodiesel RD Maximum Biodiesel Level  \
0                NaN                       NaN                        NaN   
1                NaN                       NaN                        NaN   
2                NaN                       NaN                        NaN   

  NPS Unit Name CNG Station Sells Renewable Natural Gas  \
0           NaN                                   False   
1           NaN                                     NaN   
2           NaN                                    True   

   LNG Station Sells Renewable Natural Gas  Maximum Vehicle Class  \
0                                      NaN                     MD   
1                                      NaN                     LD   
2                                      NaN                     HD   

   EV Workplace Charging  
0                    NaN  
1                    NaN  
2                    NaN  

[3 rows x 74 columns]

Number of Fuel Type code = 94369
Percentage of each value in 'Fuel Type Code':
ELEC: 87.12%
E85: 4.8%
LPG: 3.74%
BD: 1.8%
CNG: 1.62%
RD: 0.65%
HY: 0.14%
LNG: 0.13%

Question 2.	What is the number and percentage of the top 3 most common states? Make an “insight” about the location of alternative fuel centers. Where are they being deployed most?

In [ ]:
top_states = df['State'].value_counts().head(3)
total_centers = len(df)
top_states_percentage = (top_states / total_centers * 100).round(2)

In [ ]:
print("Top 3 most common states for alternative fuel centers:")
print(top_states)
print("\nPercentage of total alternative fuel centers:")
print(f"{top_states}:{top_states_percentage}%")

Top 3 most common states for alternative fuel centers:
State
CA    19402
QC     4573
NY     4478
Name: count, dtype: int64

Percentage of total alternative fuel centers:
State
CA    19402
QC     4573
NY     4478
Name: count, dtype: int64:State
CA    20.56
QC     4.85
NY     4.75
Name: count, dtype: float64%


Question 3.	Look at the column called “EV_Network.” What are the top 3 most common EV charging networks?

In [ ]:
print(df.columns)

Index(['Fuel Type Code', 'Station Name', 'Street Address',
       'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser N

From dictionary entry with the following keys:
{
    "Index(['Fuel Type'": "       'EV Level2 EVSE Num'",
    " 'Station Name'": 'EV DC Fast Count'",
 " 'Street Address'": " 'EV Other Info'",
    "": " 'EV Network'"
  }
  print(df['EV_Network'].value_counts().head(3))
  print(df['EV_Network'].value_counts(normalize=True).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).tail(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).tail(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).tail(3))
  print(df['EV_Network'].value_counts(normalize=True).sort_values(ascending=False).head(3))
  

In [ ]:
top_3_networks = df[['EV Network']].value_counts().tail(3)
print(top_3_networks)

EV Network             
AddÉnergie Technologies    2
STAY_N_CHARGE              1
HONEY_BADGER               1
Name: count, dtype: int64


4.	I want to buy a Tesla. I live in Mississippi. How many charging stations are there in my state?

In [ ]:
# Filter the DataFrame to only include Tesla charging stations
tesla_df = df[df['Fuel Type Code'] == 'TSLA']

In [ ]:
# Count the number of Tesla charging stations in Mississippi
mississippi_tesla_stations = tesla_df[tesla_df['State'] == 'MS'].shape[0]

In [ ]:
# Print the number of Tesla charging stations in Mississippi
print(f"There are {mississippi_tesla_stations} Tesla charging stations in Mississippi.")

There are 0 Tesla charging stations in Mississippi.


OR use this method to get the number of Tesla charging stations in mississippi

In [ ]:
mississippi_stations = df[df['State'] == 'Mississippi']
print(mississippi_stations.shape[0])
print(mississippi_stations.head())

0
Empty DataFrame
Columns: [Fuel Type Code, Station Name, Street Address, Intersection Directions, City, State, ZIP, Plus4, Station Phone, Status Code, Expected Date, Groups With Access Code, Access Days Time, Cards Accepted, BD Blends, NG Fill Type Code, NG PSI, EV Level1 EVSE Num, EV Level2 EVSE Num, EV DC Fast Count, EV Other Info, EV Network, EV Network Web, Geocode Status, Latitude, Longitude, Date Last Confirmed, ID, Updated At, Owner Type Code, Federal Agency ID, Federal Agency Name, Open Date, Hydrogen Status Link, NG Vehicle Class, LPG Primary, E85 Blender Pump, EV Connector Types, Country, Intersection Directions (French), Access Days Time (French), BD Blends (French), Groups With Access Code (French), Hydrogen Is Retail, Access Code, Access Detail Code, Federal Agency Code, Facility Type, CNG Dispenser Num, CNG On-Site Renewable Source, CNG Total Compression Capacity, CNG Storage Capacity, LNG On-Site Renewable Source, E85 Other Ethanol Blends, EV Pricing, EV Pricing (French

In [ ]:
#4 df_miss['ev connector types'].value_counts()
df_miss['ev_network'].value_counts()

Question 5.	What percentage of stations are owned by the Federal government?

In [ ]:
# Filter the DataFrame to only include rows with Federal government as the owner type
federal_gov_df = df[df['Owner Type Code'] == 'Federal Government']

# Calculate the percentage of Federal government-owned stations
federal_gov_percentage = (federal_gov_df.shape[0] / df.shape[0]) * 100

# Print the percentage of Federal government-owned stations
print(f"{federal_gov_percentage:.2f}% of alternative fuel stations are owned by the Federal government.")

0.00% of alternative fuel stations are owned by the Federal government.


Question 6.	What are the top 3 most popular facility types? Why do you think those facility types are the most popular?

In [ ]:
# Count the number of occurrences of each facility type
facility_type_counts = df['Facility Type'].value_counts()

# Get the top 3 most common facility types
top_3_facility_types = facility_type_counts.head(3)

# Print the top 3 most common facility types
print("Top 3 most common facility types:")
print(top_3_facility_types)

Top 3 most common facility types:
Facility Type
CONVENIENCE_STORE    4912
HOTEL                3454
CAR_DEALER           3116
Name: count, dtype: int64


In [ ]:
top_facility_types = df['Facility Type'].value_counts()
top_3_facility_types = facility_type_counts.head(3)
print(top_facility_types)

Facility Type
CONVENIENCE_STORE    4912
HOTEL                3454
CAR_DEALER           3116
GAS_STATION          2469
FUEL_RESELLER        1467
                     ... 
PHARMACY               12
WORKPLACE               6
MOTOR_POOL              5
STADIUM                 3
TNC                     1
Name: count, Length: 63, dtype: int64


Question 7. For Level 2 EV charging stations (column “EV Level2 EVSE Num”), how many Level 2 EV Charging stations have more than 10 chargers?

In [ ]:
# Filter the DataFrame to only include rows with Level 2 EV charging stations
level2_df = df[df['EV Level2 EVSE Num'] > 0]

# Count the number of Level 2 EV charging stations with more than 10 chargers
more_than_10_chargers = (level2_df['EV Level2 EVSE Num'] > 10).sum()

# Print the number of Level 2 EV charging stations with more than 10 chargers
print(f"There are {more_than_10_chargers} Level 2 EV charging stations with more than 10 chargers.")

There are 946 Level 2 EV charging stations with more than 10 chargers.


In [ ]:
num_stations_with_more_than_10_chargers = df[(df['EV Level2 EVSE Num'] > 10)].count()
print(num_stations_with_more_than_10_chargers)

Fuel Type Code                             946
Station Name                               946
Street Address                             946
Intersection Directions                     99
City                                       946
                                          ... 
NPS Unit Name                                2
CNG Station Sells Renewable Natural Gas      0
LNG Station Sells Renewable Natural Gas      0
Maximum Vehicle Class                      351
EV Workplace Charging                      945
Length: 74, dtype: int64


Question 8. List 3 questions that you would want to ask if you are looking to open a new alternative fuel center.
Below are the questions, i would ask for alternative fuel centers;

Question A: What is the demand for alternative fuel in the area?
Question B: What is the distribution of alternative fuel types in the area?
Question C: What is the distribution of facility types in the area?

In [ ]:
# Question A: What is the demand for alternative fuel in the area?
# Count the number of alternative fuel stations in the area
num_stations = df[(df['State'] == 'your_state') & (df['City'] == 'your_city')].shape[0]

# Print the number of alternative fuel stations in the area
print(f"There are currently {num_stations} alternative fuel stations in {your_city}, {your_state}.")

In [ ]:
#Question B: What is the distribution of alternative fuel types in the area?
# Count the number of occurrences of each alternative fuel type in the area
fuel_type_counts = df[(df['State'] == 'your_state') & (df['City'] == 'your_city')]['Fuel Type Code'].value_counts()

# Print the distribution of alternative fuel types in the area
print("Distribution of alternative fuel types in {your_city}, {your_state}:")
print(fuel_type_counts)

In [ ]:
# Question C: What is the distribution of facility types in the area?
# Count the number of occurrences of each facility type in the area
facility_type_counts = df[(df['State'] == 'your_state') & (df['City'] == 'your_city')]['Facility Type'].value_counts()

# Print the distribution of facility types in the area
print("Distribution of facility types in {your_city}, {your_state}:")
print(facility_type_counts)

Question 9.	What is an “insight” that you could communicate to someone who is thinking about whether to invest in a new alternative fuel center

These are the insights that i would communicate to investor to be in an alternative fuel center:

The demand for alternative fuel is growing, and there is a trend towards increased adoption of electric vehicles (EVs) in particular. According to the International Energy Agency - see reference, the number of electric cars on the road worldwide surpassed 1 million in 2020, a 40% year-on-year increase. This trend is expected to continue, with the number of electric cars projected to reach 140 million by 2030.

Investing in an alternative fuel center now could position you to take advantage of this growing market. However, it is important to carefully consider the location and type of alternative fuel center, as well as the current state of alternative fuel infrastructure in the area. By analyzing data on alternative fuel stations, such as the data available from the US Department of Energy's Alternative Fuels Data Center, you can make informed decisions about the type and location of your new alternative fuel center.

Additionally, it is important to consider the potential environmental benefits of investing in alternative fuel. Alternative fuels, such as electricity, hydrogen, and biofuels, can produce lower greenhouse gas emissions than traditional fossil fuels, which can help reduce the carbon footprint of transportation. By investing in alternative fuel, you can contribute to a more sustainable future and potentially attract customers who prioritize environmental concerns.

Overall, investing in an alternative fuel center can be a smart business decision that aligns with growing market trends and environmental goals. However, it is important to carefully consider the location, type, and potential environmental benefits of your investment.

Based on the reference below;
1. International Energy Agency. (2021). Global EV Outlook 201: Achieving an Electric Vehicle Future. Retrieved https://www.iea.org/reports/global--outlook-2021
2. US Department of Energy's Alternative Fuels Data Center. Alternative Fueling Station Counts by State and County. Retrieved from https://afdc.energy.gov/data/stations/counts
3. US Department of Energy's Alternative Fuels Data Center. Alternative Fueling Station Locations by Fuel Type. Retrieved from https://afdc.energy.gov/data/stations/stationsbyfueltype.html
4. US Environmental Protection Agency. Greenhouse Gas Emissions from Transportation. Retrieved from https://www.epa.gov/greenvehicles/fast-facts-transportation-greenhouse-gas-emissions

Question 10.	List an insight or fact about a column of your choosing. List the column name and give some piece of information about that column.

In [ ]:
# Choosing the state column
# check for missing values in the state column
print(df['State'].value_counts().head())
print("Missing values in the 'State' column: ", df['State'].isnull().sum())

State
CA    19402
QC     4573
NY     4478
ON     4379
TX     4168
Name: count, dtype: int64
Missing values in the 'State' column:  15


In [ ]:
# Counting the number of unique values in the "State" column
print("Number of unique states: ", df['State'].nunique())

Number of unique states:  65


In [ ]:
# calculate the percentage of rows for each state
state_counts = df['State'].value_counts(normalize=True)
print("Percentage of rows for each state:\n", state_counts)

Percentage of rows for each state:
 State
CA    0.205630
QC    0.048466
NY    0.047460
ON    0.046410
TX    0.044174
        ...   
AK    0.000699
YT    0.000435
PR    0.000233
NT    0.000032
KA    0.000011
Name: proportion, Length: 65, dtype: float64
